src/services/ghost/GhostStreamingParser.ts

```diff
@@ -0,0 +1,510 @@
+import * as vscode from "vscode"
+import { structuredPatch } from "diff"
+import { GhostSuggestionContext, GhostSuggestionEditOperationType } from "./types"
+import { GhostSuggestionsState } from "./GhostSuggestions"
+
+// Special marker for cursor position
+const CURSOR_MARKER = "<<<AUTOCOMPLETE_HERE>>>"
+
+export interface StreamingParseResult {
+	suggestions: GhostSuggestionsState
+	isComplete: boolean
+	hasNewSuggestions: boolean
+}
+
+export interface ParsedChange {
+	search: string
+	replace: string
+	cursorPosition?: number // Offset within replace content where cursor should be positioned
+}
+
+/**
+ * Streaming XML parser for Ghost suggestions that can process incomplete responses
+ * and emit suggestions as soon as complete <change> blocks are available
+ */
+export class GhostStreamingParser {
+	private buffer: string = ""
+	private completedChanges: ParsedChange[] = []
+	private lastProcessedIndex: number = 0
+	private context: GhostSuggestionContext | null = null
+	private streamFinished: boolean = false
+
+	constructor() {}
+
+	/**
+	 * Initialize the parser with context
+	 */
+	public initialize(context: GhostSuggestionContext): void {
+		this.context = context
+		this.reset()
+	}
+
+	/**
+	 * Reset parser state for a new parsing session
+	 */
+	public reset(): void {
+		this.buffer = ""
+		this.completedChanges = []
+		this.lastProcessedIndex = 0
+		this.streamFinished = false
+	}
+
+	/**
+	 * Process a new chunk of text and return any newly completed suggestions
+	 */
+	public processChunk(chunk: string): StreamingParseResult {
+		if (!this.context) {
+			throw new Error("Parser not initialized. Call initialize() first.")
+		}
+
+		// Add chunk to buffer
+		this.buffer += chunk
+
+		// Extract any newly completed changes from the current buffer
+		const newChanges = this.extractCompletedChanges()
+
+		let hasNewSuggestions = newChanges.length > 0
+
+		// Add new changes to our completed list
+		this.completedChanges.push(...newChanges)
+
+		// Check if the response appears complete
+		let isComplete = this.isResponseComplete()
+
+		// Apply very conservative sanitization only when the stream is finished
+		// and we still have no completed changes but have content in the buffer
+		if (this.completedChanges.length === 0 && this.buffer.trim().length > 0 && this.streamFinished) {
+			const sanitizedBuffer = this.sanitizeXMLConservative(this.buffer)
+			if (sanitizedBuffer !== this.buffer) {
+				// Re-process with sanitized buffer
+				this.buffer = sanitizedBuffer
```

Comment:
I'm curious how often this happens? It might be nice to get telemetry on it eventually!